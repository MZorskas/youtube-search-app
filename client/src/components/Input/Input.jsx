import React from 'react';
import './Input.scss';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const Input = ({
  className,
  onChange,
  placeholder,
  validationError,
  id,
  label,
  autoFocus,
}) => {
  return (
    <React.Fragment>
      <div className="input-wrapper" data-testid="input">
        {label && (
          <label className="input-wrapper__label" htmlFor={id}>
            {label}
          </label>
        )}
        <input
          className={classNames(
            'input-wrapper__input',
            {
              'input-wrapper__input--error': validationError,
            },
            className
          )}
          id={id}
          onChange={onChange}
          placeholder={placeholder}
          autoFocus={autoFocus}
        />
      </div>
    </React.Fragment>
  );
};

Input.propTypes = {
  className: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  placeHolder: PropTypes.string,
  validationError: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  autoFocus: PropTypes.bool,
};

Input.defaultProps = {
  placeholder: 'Gifs',
};

export default Input;
