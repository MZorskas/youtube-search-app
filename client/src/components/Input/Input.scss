@import 'src/styles/mixins/index';
@import 'src/styles/base/base';

.input-wrapper {
  display: flex;
  width: 100%;

  &__input {
    @include text-style-h3;

    border: none;
    box-sizing: border-box;
    margin: 0;

    &--error {
      background-color: lighten($color-error-primary, 20%);
    }
  }

  &__label {
    @include visually-hidden;
  }
}
