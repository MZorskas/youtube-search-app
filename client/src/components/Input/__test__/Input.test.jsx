import React from 'react';
import Input from '../Input';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const mockOnChange = jest.fn();

const inputSetup = () => {
  const utils = render(
    <Input
      onChange={mockOnChange}
      className="input-test"
      placeholder="Videos"
      id="search-input"
      label="Search"
      autoFocus
    />
  );
  const input = screen.getByPlaceholderText('Videos');
  return {
    input,
    ...utils,
  };
};

describe('Input component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');

    expect(screen.queryByTestId('input')).not.toBeInTheDocument();
    render(<Input />);
    expect(screen.queryByTestId('input')).toBeInTheDocument();

    expect(spy).toHaveBeenCalled();
  });

  it('renders without propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');

    expect(screen.queryByTestId('input')).not.toBeInTheDocument();
    inputSetup();
    expect(screen.queryByTestId('input')).toBeInTheDocument();

    expect(spy).not.toHaveBeenCalled();
  });

  // it('renders and autofocus inputField', () => {
  //   expect(screen.queryByTestId('input')).not.toBeInTheDocument();
  //   render(<Input onChange={mockOnChange} placeholder="Videos" />);
  //   expect(screen.queryByTestId('input')).toBeInTheDocument();

  //   expect(screen.getByPlaceholderText('Videos')).toHaveFocus();
  // });

  it('renders with additional className', () => {
    expect(screen.queryByTestId('input')).not.toBeInTheDocument();
    render(
      <Input
        onChange={mockOnChange}
        className="input-test"
        placeholder="Videos"
      />
    );
    expect(screen.getByPlaceholderText('Videos').className).toBe(
      'input-wrapper__input input-test'
    );
  });

  it('updates input value', () => {
    const { input } = inputSetup();

    fireEvent.change(input, { target: { value: '123' } });
    expect(input.value).toBe('123');
  });

  it('removes input value', () => {
    const { input } = inputSetup();

    fireEvent.change(input, { target: { value: '123' } });
    expect(input.value).toBe('123');
    fireEvent.change(input, { target: { value: '' } });
    expect(input.value).toBe('');
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Input
          onChange={mockOnChange}
          className="input-test"
          placeholder="Videos"
          id="search-input"
          label="Search"
          autoFocus
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
