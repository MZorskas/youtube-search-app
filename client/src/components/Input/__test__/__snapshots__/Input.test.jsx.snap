// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component matches snapshot 1`] = `
<div
  className="input-wrapper"
  data-testid="input"
>
  <label
    className="input-wrapper__label"
    htmlFor="search-input"
  >
    Search
  </label>
  <input
    autoFocus={true}
    className="input-wrapper__input input-test"
    id="search-input"
    onChange={[MockFunction]}
    placeholder="Videos"
  />
</div>
`;
